src/visibility.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright 2020-2023 the Deno authors. All rights reserved. MIT license.
+
+use crate::util::graph::resolve_deno_graph_module;
+use crate::util::symbol::get_module_info;
+use crate::util::symbol::symbol_has_ignorable_js_doc_tag;
+use crate::DocError;
+
+use deno_ast::SourceRange;
+use deno_graph::symbols::DefinitionPath;
+use deno_graph::symbols::ResolvedSymbolDepEntry;
+use deno_graph::symbols::SymbolDecl;
+use deno_graph::symbols::UniqueSymbolId;
+use deno_graph::ModuleGraph;
+use indexmap::IndexMap;
+use indexmap::IndexSet;
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+pub struct SymbolDeclDeps {
+  pub symbol_id: UniqueSymbolId,
+  pub decl_range: SourceRange,
+  pub deps: IndexSet<UniqueSymbolId>,
+  /// If the path to this declaration had an ignorable js doc tag.
+  pub had_ignorable_tag: bool,
+}
+
+#[derive(Default)]
+pub struct SymbolDeps(IndexMap<SourceRange, Vec<SymbolDeclDeps>>);
+
+impl SymbolDeps {
+  pub fn is_empty(&self) -> bool {
+    self.0.is_empty()
+  }
+
+  pub fn iter(&self) -> impl Iterator<Item = &SymbolDeclDeps> {
+    self.0.values().flatten()
+  }
+
+  fn add(
+    &mut self,
+    symbol_id: UniqueSymbolId,
+    symbol_decl: &SymbolDecl,
+    dep_id: UniqueSymbolId,
+    had_ignorable_tag: bool,
+  ) {
+    self
+      .0
+      .entry(symbol_decl.range)
+      .or_default()
+      .push(SymbolDeclDeps {
+        symbol_id,
+        decl_range: symbol_decl.range,
+        deps: IndexSet::from([dep_id]),
+        had_ignorable_tag,
+      });
+  }
+}
+
+pub struct SymbolVisibility {
+  root_exported_ids: HashMap<UniqueSymbolId, SymbolDeps>,
+  /// Symbol identifiers that are not exported, but are referenced
+  /// by exported symbols.
+  non_exported_public_ids: HashSet<UniqueSymbolId>,
+}
+
+impl SymbolVisibility {
+  pub fn build(
+    graph: &ModuleGraph,
+    root_symbol: &deno_graph::symbols::RootSymbol,
+  ) -> Result<Self, DocError> {
+    // get all the symbols that are exported from the roots of the graph
+    let mut root_exported_ids = analyze_root_exported_ids(graph, root_symbol)?;
+
+    // now analyze for all the non-exported types that are referenced by exported types
+    // along with filling in any non-exported dependencies of root exported types
+    let mut non_exported_public_ids = HashSet::new();
+    let mut pending_symbol_ids =
+      root_exported_ids.keys().copied().collect::<Vec<_>>();
+    while let Some(original_id) = pending_symbol_ids.pop() {
+      let module_info = root_symbol
+        .get_module_from_id(original_id.module_id)
+        .unwrap();
+      let symbol = module_info.symbol(original_id.symbol_id).unwrap();
+      let decl_deps = symbol
+        .decls()
+        .iter()
+        .filter(|d| !d.has_overloads())
+        .flat_map(|decl| decl.deps().into_iter().map(move |dep| (decl, dep)))
+        .map(|(decl, dep)| (symbol, decl, dep))
+        .chain(
+          symbol
+            .members()
+            .iter()
```

Comment:
This function is really complex, would it make sense to unit test it?

Reply:
It's captured by the integration tests.